chain/tsresolver/tipset_resolver.go

```diff
@@ -0,0 +1,164 @@
+package tsresolver
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	logging "github.com/ipfs/go-log/v2"
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/go-f3/certs"
+	"github.com/filecoin-project/go-state-types/abi"
+
+	"github.com/filecoin-project/lotus/api"
+	"github.com/filecoin-project/lotus/chain/actors/policy"
+	"github.com/filecoin-project/lotus/chain/types"
+	"github.com/filecoin-project/lotus/chain/types/ethtypes"
+)
+
+var log = logging.Logger("chain/tsresolver")
+
+const (
+	EthBlockSelectorEarliest  = "earliest"
+	EthBlockSelectorPending   = "pending"
+	EthBlockSelectorLatest    = "latest"
+	EthBlockSelectorSafe      = "safe"
+	EthBlockSelectorFinalized = "finalized"
+)
+
+type TipSetLoader interface {
+	GetHeaviestTipSet() (ts *types.TipSet)
+	LoadTipSet(ctx context.Context, tsk types.TipSetKey) (*types.TipSet, error)
+	GetTipsetByHeight(ctx context.Context, h abi.ChainEpoch, anchor *types.TipSet, prev bool) (*types.TipSet, error)
+}
+
+type F3 interface {
+	GetLatestCert(ctx context.Context) (*certs.FinalityCertificate, error)
+	IsEnabled() bool
+	IsRunning() (bool, error)
+}
+
+type TipSetResolver interface {
+	ResolveEthBlockSelector(ctx context.Context, selector string, strict bool) (*types.TipSet, error)
+}
+
+type tipSetResolver struct {
+	loader TipSetLoader
+	f3     F3
+}
+
+var _ TipSetResolver = (*tipSetResolver)(nil)
+
+func NewTipSetResolver(loader TipSetLoader, f3 F3) TipSetResolver {
+	return &tipSetResolver{
+		loader: loader,
+		f3:     f3,
+	}
+}
+
+// ResolveEthBlockSelector resolves an Ethereum block selector string to a TipSet.
+//
+// The selector can be one of:
+//   - "pending": the chain head
+//   - "latest": the TipSet with the latest executed messages (head - 1)
+//   - "safe": the TipSet with messages executed at least eth.SafeEpochDelay (30) epochs ago or the
+//     latest F3 finalized TipSet
+//   - "finalized": the TipSet with messages executed at least policy.ChainFinality (900) epochs ago
+//     or the latest F3 finalized TipSet
+//   - a decimal block number: the TipSet at the given height
+//   - a 0x-prefixed hex block number: the TipSet at the given height
+//
+// If a specific block number is specified and `strict` is true, an error is returned if the block
+// number resolves to a null round, otherwise in the case of a null round the first non-null TipSet
+// immediately before the null round is returned.
+func (tsr *tipSetResolver) ResolveEthBlockSelector(ctx context.Context, selector string, strict bool) (*types.TipSet, error) {
```

Comment:
I'm not sure I agree with this since it's not a closed set of possible values: it's an awkward union of a set of fixed strings (defined as constants in here for clarity) and the various string forms that can be decoded by `ethtypes.EthUint64`. This method is more like a "figure out what this string I've been given means" rather than "take one of these and transform it into one of those". Can you talk more about how you think it would help by defining it as a type?

Reply:
> "figure out what this string I've been given means"

I take this function to do more than that because it is exported? But I see the type is not exported. 

> how you think it would help by defining it as a type?

The function clearly looks for well known string values. Using a type is an executable definition of that intent for me as a reader of code.

If this not exported anywhere then this is bikeshedding. Marking as resolved as it is a non blocker for me and a matter of oppinion.

Reply:
it is exported on the interface and it's used as exported

I think there's a refactor here that will align this better, we have a `EthBlockNumberOrHash` that _should_ have served this purpose but it looks like it wasn't properly integrated when `"finalized"` was originally introduced, so repurposing it for this is a bit awkward and may break some things; I don't want to do that right now because it's not at all simple.

Opened an issue describing the shape of the problem here: https://github.com/filecoin-project/lotus/issues/12766

Thanks for raising.